### PR TITLE
Fix workflow-state alternate run-dir

### LINKED ISSUE #6030: workflow_state xtrigger and CLI command do not work across users
### Description

```
>>> import workflow_state
>>> from workflow_state import workflow_state as ws
>>> ws('a', 'b', '20240101T1200Z', cylc_run_dir='/home/another_user')
Traceback (most recent call last):
  File "/home/user/miniconda3/envs/cylc-dev/lib/python3.11/site-packages/cylc/flow/workflow_files.py", line 872, in infer_latest_run
    id_ = str(path.relative_to(cylc_run_dir))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/miniconda3/envs/cylc-dev/lib/python3.11/pathlib.py", line 730, in relative_to
    raise ValueError("{!r} is not in the subpath of {!r}"
ValueError: '/home/another_user/a' is not in the subpath of '/home/548/tc2375/cylc-run' OR one path is relative and the other is absolute.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/548/tc2375/cylc/cylc/flow/xtriggers/workflow_state.py", line 87, in workflow_state
    _, workflow = infer_latest_run(Path(cylc_run_dir, workflow))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/miniconda3/envs/cylc-dev/lib/python3.11/site-packages/cylc/flow/workflow_files.py", line 874, in infer_latest_run
    raise ValueError(f"{path} is not in the cylc-run directory")
ValueError: /home/another_user/a is not in the cylc-run directory



>>> from pathlib import Path
>>> a=Path('/home/tc2375')
>>> a.relative_to('/home/another_user')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/miniconda3/envs/cylc-dev/lib/python3.11/pathlib.py", line 730, in relative_to
    raise ValueError("{!r} is not in the subpath of {!r}"
ValueError: '/home/tc2375' is not in the subpath of '/home/another_user' OR one path is relative and the other is absolute.
>>>
```

### Reproducible Example

See above. The `relative_to` method does not work for across user checking. 
You can do similar from the command line interface, but make sure to use the `--debug` flag otherwise the error message is misleading/unhelpful.

### Expected Behaviour

Should be able to use workflow_state across different users.

### PR DESCRIPTION
**Pull request by hjoliver**
Close #6030

~Targeting 8.3.0 because it seems we're unlikely to get another 8.2 release out before then, and 8.2.x has some conflicting changes.~ (Can rebase to 8.2.x if really necessary <--- done).

TODO:
- [x] CLI: `cylc workflow-state`
- [x] `workflow_state` xtrigger

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/workflow_files.py [hunk 1]
```diff
@@ -862,16 +862,22 @@ def check_reserved_dir_names(name: Union[Path, str]) -> None:
...
 
 
-def infer_latest_run_from_id(workflow_id: str) -> str:
-    run_dir = Path(get_workflow_run_dir(workflow_id))
-    _, id_ = infer_latest_run(run_dir)
+def infer_latest_run_from_id(
+    workflow_id: str, alt_run_dir: Optional[str] = None
+) -> str:
+    if alt_run_dir is not None:
+        run_dir = Path(alt_run_dir) / Path(workflow_id)
```
Comment (dwsutherland) [inline cylc/flow/workflow_files.py:869]: Can you do this with `Path`?

Comment (dwsutherland) [reply cylc/flow/workflow_files.py:869]: Actually yes, cool..

**Review by dwsutherland — APPROVED** [review]
LGTM 👍

File: cylc/flow/workflow_files.py (hunk 1 above)
Comment (MetRonnie) [reply cylc/flow/workflow_files.py:869]: More usual to do this though
```suggestion
        run_dir = Path(alt_run_dir, workflow_id)
```

File: tests/functional/workflow-state/08-other-run-dir.t [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#-------------------------------------------------------------------------------
+# Test cylc workflow-state "template" option
```
Comment (dwsutherland) [inline tests/functional/workflow-state/08-other-run-dir.t:18]: Oh, I should have changed this comment.. Apologies

Comment (hjoliver) [inline tests/functional/workflow-state/08-other-run-dir.t:18]:
```suggestion
# Test cylc workflow-state for alternate run dir.
```

**Comment by hjoliver** [comment]
Rebased onto 8.2.x branch for 8.2.5

**Comment by hjoliver** [comment]
@dwsutherland  and @MetRonnie  - I've managed to simplify the code change quite a bit, and replaced the functional test with unit tests.  Hoping we can bang this into 8.2.5.

**Review by MetRonnie — APPROVED** [review]

File: cylc/flow/scripts/workflow_state.py [hunk 3]
```diff
@@ -231,15 +227,11 @@ def main(parser: COP, options: 'Values', workflow_id: str) -> None:
...
-    if options.run_dir:
-        run_dir = expand_path(options.run_dir)
-    else:
-        run_dir = get_cylc_run_dir()
+    workflow_id = infer_latest_run_from_id(workflow_id, options.alt_run_dir)
 
     pollargs = {
         'workflow_id': workflow_id,
-        'run_dir': run_dir,
+        'run_dir': expand_path(options.alt_run_dir or get_cylc_run_dir()),
```
Comment (dwsutherland) [inline cylc/flow/scripts/workflow_state.py:234]: Doesn't `get_cylc_run_dir` already employ `expand_path`?
This is why I just included an optional arg to `get_cylc_run_dir`, and removed the import of `expand_path` in a couple of places.

File: cylc/flow/xtriggers/workflow_state.py [hunk 4]
```diff
@@ -78,13 +76,12 @@ def workflow_state(
             to this xtrigger.
 
     """
-    if cylc_run_dir:
-        cylc_run_dir = expand_path(cylc_run_dir)
-    else:
-        cylc_run_dir = get_cylc_run_dir()
+    workflow = infer_latest_run_from_id(workflow, cylc_run_dir)
+    cylc_run_dir = expand_path(cylc_run_dir or get_cylc_run_dir())
```
Comment (dwsutherland) [inline cylc/flow/xtriggers/workflow_state.py:80]: same here

**Review by dwsutherland — APPROVED** [review]
Ok, can probably the point below

File: cylc/flow/scripts/workflow_state.py (hunk 3 above)
Comment (hjoliver) [reply cylc/flow/scripts/workflow_state.py:234]: Yeah, all it does is expand the default run-dir.  I'm considering other options. The various workflow run path functions are getting a bit convoluted...

**Review by dwsutherland — APPROVED** [review]
Yeah, improvement on my attempt 👍

**Comment by hjoliver** [comment]
Thanks, I'll merge this if tests pass...

**Pull request merged**